### ISSUE TITLE
Using numpy.ndarray for Input Data

### ISSUE THREAD

**Issue by II-Matto** [dmlc/mxnet#199]
After trying the MNIST `mlp.py` example, I wanted to use my own data to train an MLP. So I load input from `*.txt` files in python as numpy.ndarray, which are then taken as arguments of `model.fit()`. But I keep getting the following error:

> Traceback (most recent call last):
  File "test.py", line 36, in <module>
    model.fit(X=train_data, y=train_label, eval_data=(val_data, val_label))
  File "/home/code/xxx/mxnet/mxnet_master/python/mxnet/model.py", line 675, in fit
    logger=logger)
  File "/home/code/xxx/mxnet/mxnet_master/python/mxnet/model.py", line 273, in _train_multi_device
    label[islice].copyto(target)
  File "/home/code/xxx/mxnet/mxnet_master/python/mxnet/ndarray.py", line 320, in copyto
    return NDArray._copyto(self, out=other)
  File "/home/code/xxx/mxnet/mxnet_master/python/mxnet/ndarray.py", line 640, in generic_ndarray_function
    c_array(NDArrayHandle, [v.handle for v in mutate_vars])))
  File "/home/code/xxx/mxnet/mxnet_master/python/mxnet/base.py", line 72, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: [21:58:27] src/ndarray/ndarray.cc:158: Check failed: from.shape() == to->shape() operands shape mismatch

My data is represented as 1500-dim feature vectors. The shapes of my data array are:
> train_data.shape = (1800, 1500)
train_label.shape = (1800, 1)
val_data.shape = (200, 1500)
val_label.shape = (200, 1)

The MLP has two output units for the last layer. And I have tried transposing the input data array, using 1500 units for `fc1`, and using 2-dim labels (for two output units). None of them helps.

So what is the correct way to use my own data? Thanks.

**Comment by antinucleon** [comment]
Please check https://github.com/dmlc/mxnet/blob/master/example/mnist/mlp_numpy.py

**Comment by tqchen** [comment]
I guess the current problem is the label shape should be 1D ndarray, while @II-Matto 's usecase it was shaped as 2D. Maybe we should update the code to make it compatible with both @antinucleon

**Comment by II-Matto** [comment]
@antinucleon Thanks for you example! I have successfully trained an MLP with my own data.

I noticed that when using numpy.ndarray as input, the validation set is not used. I tried myself but it reported error. To be specific, I have tried two ways. One is
```
model.fit(X=train_data, y=train_label, eval_data=(val_data, val_label))
```
This results in the error of "too many values to unpack" (line 325 in model.py). I checked the model.py and googled for a while. Then I tried the other way.
```
val_set = [(v, c) for v, c in zip(val_data, val_label)]
model.fit(X=train_data, y=train_label, eval_data=val_set)
```
But it results in the error of "invalid index to scalar variable" (line 328 in model.py). It seems to be that the unpacked `label` is a scalar so the index `islice` is not applicable.

So what is the right way to use numpy.ndarray for the `eval_data` argument?

**Comment by tqchen** [comment]
Sorry for delayed reply. Seems we have not yet support using numpy in validation. This however, should be readily supported by taking a tuple of data, label as you suggested. 

We can easily convert the data, label tuple into a DataIter wanted by this funciton https://github.com/dmlc/mxnet/blob/master/python/mxnet/model.py#L648

@II-Matto Are you interested in open a PR for this?

**Comment by II-Matto** [comment]
@tqchen Sure, I will do it tomorrow!

**Comment by II-Matto** [comment]
@tqchen I have opened a PR for supporting `eval_data` as ndarray pair. I have tested the code with my own data, and it works fine. Hope I didn't make anything wrong.

**Comment by tqchen** [comment]
fixed by #278
